### PR TITLE
support destructuring params with defaults.

### LINKED ISSUE #81: Default values for named parameters ({x} = {}) incorrectly handled (angular2/src/core/annotations/annotations.dart)

### PR DESCRIPTION
**Pull request by alexeagle**
Fixes #81

### CODE REVIEW

File: main.ts [hunk 1]
```diff
@@ -699,21 +699,23 @@ class Translator {
...
             // TODO(martinprobst): These are currently silently ignored.
             // this.reportError(paramDecl.type, 'types on named parameters are unsupported');
           }
-        }
-        this.visit(paramDecl.name);
-        if (paramDecl.initializer) {
-          this.emit('=');
-          this.visit(paramDecl.initializer);
-          this.emit(']');
+          this.visit(paramDecl.name);
```
Comment (mprobst) [inline main.ts]: I think this code should check that the initialiser is just an empty object assignment, and if not, report an error.

File: main.ts [hunk 2]
```diff
@@ -699,21 +699,23 @@ class Translator {
       case ts.SyntaxKind.Parameter:
         var paramDecl = <ts.ParameterDeclaration>node;
         if (paramDecl.dotDotDotToken) this.reportError(node, 'rest parameters are unsupported');
-        if (paramDecl.initializer) this.emit('[');
-        this.visitEachIfPresent(paramDecl.decorators);
-        if (paramDecl.type) {
-          if (paramDecl.name.kind !== ts.SyntaxKind.ObjectBindingPattern) {
-            this.visit(paramDecl.type);
-          } else {
+        if (paramDecl.name.kind === ts.SyntaxKind.ObjectBindingPattern) {
```
Comment (mprobst) [inline main.ts]: This is sort of beyond the threshold of a single case block, isn't it? Maybe extract to a separate function?

File: test/DartTest.ts [hunk 3]
```diff
@@ -206,12 +206,16 @@ describe('transpile to dart', () => {
       expectErroneousCode('function x({a}: number) { return a + b; }')
           .to.throw('types on named parameters are unsupported');
     });
+    it('translates destructuring parameters', () => {
+      expectTranslate('function x({p = null, d = false} = {}) {}')
```
Comment (mprobst) [inline test/DartTest.ts]: add a test to fail if the initializer is more than just an empty object?

File: main.ts (hunk 1 above)
Comment (alexeagle) [inline main.ts]: yes, good idea

File: main.ts (hunk 2 above)
Comment (alexeagle) [inline main.ts]: extracted a function for the unusual case.
This file is getting unmanageable, the distance from this point to the helper function is large. don't have a suggestion what to do about it...

File: test/DartTest.ts (hunk 3 above)
Comment (alexeagle) [inline test/DartTest.ts]: done

File: main.ts [hunk 4]
```diff
@@ -122,6 +122,23 @@ class Translator {
...
+      // TODO(martinprobst): These are currently silently ignored.
+      // this.reportError(paramDecl.type, 'types on named parameters are unsupported');
+    }
+    this.visit(paramDecl.name);
+    if (paramDecl.initializer) {
+      if (paramDecl.initializer.kind !== ts.SyntaxKind.ObjectLiteralExpression) {
+        this.reportError(paramDecl, 'named parameter initializer must be empty object literal');
+      }
+      if ((<ts.ObjectLiteralExpression>paramDecl.initializer).properties.length > 0) {
+        this.reportError(paramDecl, 'non-empty named parameter initializers are unsupported');
```
Comment (mprobst) [inline main.ts]: I'd say the same as above, `initializers for named parameters must be empty object literals`, fewer negations and tells you what to do.

File: test/DartTest.ts [hunk 5]
```diff
@@ -165,27 +164,21 @@ describe('transpile to dart', () => {
...
-    it('translates on functions', () => {
-      expectTranslate('@A function f() {}').to.equal(' @ A f ( ) { }');
-    });
-    it('translates on properties', () => {
-      expectTranslate('class X { @A p; }').to.equal(' class X { @ A var p ; }');
-    });
-    it('translates on parameters', () => {
-      expectTranslate('function f (@A p) {}').to.equal(' f ( @ A p ) { }');
-    });
+    it('translates plain decorators',
```
Comment (mprobst) [inline test/DartTest.ts]: could you split out these changes? It's kind of hard to figure out what actually changed here.

File: main.ts [hunk 6]
```diff
@@ -122,6 +122,23 @@ class Translator {
...
   }
 
+  visitNamedParameter(paramDecl: ts.ParameterDeclaration) {
+    this.visitEachIfPresent(paramDecl.decorators);
+    if (paramDecl.type) {
+      // TODO(martinprobst): These are currently silently ignored.
+      // this.reportError(paramDecl.type, 'types on named parameters are unsupported');
+    }
+    this.visit(paramDecl.name);
+    if (paramDecl.initializer) {
```
Comment (mprobst) [inline main.ts]: actually that's going sort of towards a linter, but should this also fail if there is no initializer for a named parameter list? That's AFAICT always wrong (fails if you don't pass any named parameters).

File: main.ts (hunk 4 above)
Comment (alexeagle) [inline main.ts]: done, and added another test assertion

File: test/DartTest.ts (hunk 5 above)
Comment (alexeagle) [inline test/DartTest.ts]: hey, no fair! if I run clang-format and get spurious deltas, that's the fault of whoever commited previously.
I'll send a format-only change and rebase this on top.

File: main.ts (hunk 6 above)
Comment (alexeagle) [inline main.ts]: I think you're saying there's another validation we could add? I didn't quite follow. Maybe you're saying there's a bug in this code?

**Comment by mprobst** [comment]
LGTM.
